proposals/0000-pure-template-haskell.md

```diff
@@ -0,0 +1,294 @@
+- Once much of the ecosystem is `Wimpure-splices`-free, switch `EnforceTemplateHaskellPurity` on by default.
+
+In parallel, as soon as there is significant explicit usage of `ImpureTemplateHaskell` in the wild, this will lead to pressure on library authors to purify the Template Haskell functions that they provide (since Haskellers will hate having something saying “impure” at the top of their files).
+
+This in turn will lead to pressure to ensure that `PureQ` has enough capabilities to cover major use cases (see “Supporting sufficiently many use cases” below).
+
+However, this process does not block changing the defaults, and can continue afterwards.
+
+## Examples
+
+Examples of `EnforceTemplateHaskellPurity`. 
+We're ignoring the stage restriction here for compactness, since it’s orthogonal to this proposal.
+
+```
+-- GOOD: Generating syntax definitions is entirely pure, but typically requires Reify
+$makeLenses ‘Foo
+
+-- BAD: reading a file requires IO, ‘embedFile’ has type ‘FilePath -> Q Exp’
+myfile1 = $(embedFile "dirName/fileName")
+-- GOOD: explicitly uses an impure splice
+myfile2 = $!(embedFile “dirName/fileName”)
+
+-- BAD: the inner splice is impure, leading to a MonadIO constraint on the overall splice, 
+-- which cannot be satisfied by a top-level pure splice
+myLazyfile3 = $(BSL.toLazy $(embedFile “dirName/fileName”))
+-- GOOD: explicitly uses an impure splice
+myLazyfile3 = $!(BSL.toLazy $(embedFile “dirName/fileName”))
+
+-- this is an impure quote
+q :: (MonadIO m, Quote m) => m Exp
+q = [| $(liftIO (putStrLn "hello") >> pure (LitE (LitInteger 5))) |]
+
+-- BAD: the quote is impure, so we need an impure splice
+myInt1 = $q
+-- BAD: use an impure splice
+myInt2 = $!q
+```
+
+## Effect and Interactions
+
+### Direct effects
+
+The effect of `EnforceTemplateHaskellPurity` is to make the evaluation of top-level normal splices pure. 
+We think it is justified to move towards having this as the default, seeing as we expect it to cover the majority of cases (and it fits with Haskell’s general ethos of pure-first).
+
+`ImpureTemplateHaskell` then allows top-level splices that need IO so long as they use the new syntax for impure splices.
+
+The overall effect is that the programmer can statically assert that execution of Template Haskell will not perform IO at compile time (modulo `unsafePerformIO`). 
+However it leaves the option for them to opt into this explicitly at key points.
+
+### unsafePerformIO
+
+How can we claim that we are gaining any safety when `unsafePerformIO` exists? 
+Surely this allows users to smuggle IO into anywhere! 
+So the compiler cannot assume that there is no IO (and nor can the user!).
+
+This is true, but it is true of normal Haskell code as well. 
+The approach we usually take is to say “the compiler is entitled to assume that there is no IO if it’s not in the type; if you use unsafePerformIO then unpredictable things may happen and that’s your problem”. 
```

Comment:
But if you don't care about `unsafePerformIO`, what is the motivation in the first place?

It's not like users who use `embedFile` today don't *know* that doing so makes their build impure. It's just that they don't have a proper alternative, one that works when GHC, HLS, Cabal or Stack drives the build.